sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,442 @@
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toShort
+                toColumn.appendShorts(batchSize, data)
+              case IntegerType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+              case LongType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0)
+                toColumn.appendLongs(batchSize, data)
+
+              case DateType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+
+              case TimestampType =>
+                val data = fromColumn.asInstanceOf[TimestampColumnVector]
+                toColumn.appendLongs(batchSize, data.time(0) * 1000L + data.nanos(0) / 1000L)
+
+              case FloatType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0).toFloat
+                toColumn.appendFloats(batchSize, data)
+              case DoubleType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0)
+                toColumn.appendDoubles(batchSize, data)
+
+              case StringType =>
+                val data = fromColumn.asInstanceOf[BytesColumnVector]
+                var index = 0
+                while (index < batchSize) {
+                  toColumn.appendByteArray(data.vector(0), data.start(0), data.length(0))
+                  index += 1
+                }
+              case BinaryType =>
+                val data = fromColumn.asInstanceOf[BytesColumnVector]
+                var index = 0
+                while (index < batchSize) {
+                  toColumn.appendByteArray(data.vector(0), data.start(0), data.length(0))
+                  index += 1
+                }
+
+              case DecimalType.Fixed(precision, scale) =>
+                val d = fromColumn.asInstanceOf[DecimalColumnVector].vector(0)
+                val value = Decimal(d.getHiveDecimal.bigDecimalValue, d.precision(), d.scale)
+                value.changePrecision(precision, scale)
+                if (precision <= Decimal.MAX_INT_DIGITS) {
+                  toColumn.appendInts(batchSize, value.toUnscaledLong.toInt)
+                } else if (precision <= Decimal.MAX_LONG_DIGITS) {
+                  toColumn.appendLongs(batchSize, value.toUnscaledLong)
+                } else {
+                  val bytes = value.toJavaBigDecimal.unscaledValue.toByteArray
+                  var index = 0
+                  while (index < batchSize) {
+                    toColumn.appendByteArray(bytes, 0, bytes.length)
+                    index += 1
+                  }
+                }
+
+              case dt =>
+                throw new UnsupportedOperationException(s"Unsupported Data Type: $dt")
+            }
+          }
+        } else if (!field.nullable || fromColumn.noNulls) {
```

Comment:
I'm not sure I get how this if stmt works right.

Reply:
Could you elaborate your concern on this?

Reply:
While I get your schema says the column has no nulls (!field.nullable) that isn't a guarantee that the column returned by the reader is noNulls.  There should be an assertion or perhaps better yet an exception thrown for a data error.  (My other comment about selectedInUse probably ought to be a data error, too).

This method talks about ORC reader -- yet really this is general Hive to Spark vectorization code and it will in one way or another be generalized and reused.  In Hive, there is work to vectorize Parquet fully.  We also support vectorized reading of text files and through LLAP.

Reply:
(Nulls can sneak in when enforcing decimal precision/scale, for example).

Reply:
Throwing data error exceptions can be nasty -- perhaps setting the value to a known value and a warning might be better.  I noticed leveraging of not nullable in the Hive optimizer recently but we will see if there are any practical issues that arise...

Reply:
Oh, I missed your comment here, @mmccline . Thank you for review!

Reply:
@mmccline . For this, I'll respect only `fromColumn.noNulls` here.